lib/spack/spack/variant.py

```diff
@@ -598,6 +605,164 @@ def substitute_abstract_variants(spec):
         spec.variants.substitute(new_variant)
 
 
+# The class below inherit from Sequence to disguise as a tuple and comply
+# with the semantic expected by the 'values' argument of the variant directive
+class DisjointSetsOfValues(Sequence):
+    """Allows combinations from one of many mutually exclusive sets.
+
+    The value ``('none',)`` is reserved to denote the empty set
+    and therefore no other set can contain the item ``'none'``.
```

Comment:
> It is a strange effect that I just noticed that when 'none' is allowed, it doesn't appear in values: that's worth mentioning.

Along with the bug you found above, this is another effect I didn't notice in my previous implementation (and for which there was no test). To explain my previous choice: I tried to implement the variant object in such a way that I didn't have to modify `AutotoolsPackage`. 

Since now a modification to that class seems necessary to fix the 2 issues you raised, I went with a variation of your first suggestion:
https://github.com/spack/spack/blob/3cc558f9d237437bd725ed7ae9ea5bd12bd5df3f/lib/spack/spack/build_systems/autotools.py#L362-L374

by allowing a variant object to expose an attribute called `feature_values`. This restores the expected semantic of a tuple, brings back the possibility to ask for `foo=none` even when `none` is not the default and shows `none` in the list of values.